scipy/optimize/zeros.py

```diff
@@ -293,11 +306,13 @@ class of similar problems can be solved together.
                 p = (p1 + p0) / 2.0
                 return _results_select(full_output, (p, funcalls, itr + 1, _ECONVERGED))
             else:
-                p = p1 - q1 * (p1 - p0) / (q1 - q0)
-            if abs(p - p1) < tol:
+                if abs(q1) > abs(q0):
+                    p = (-q0 / q1 * p1 + p0) / (1 - q0 / q1)
+                else:
+                    p = (-q1 / q0 * p0 + p1) / (1 - q1 / q0)
```

Comment:
I apologize for my ignorance, but I don't understand the difference between these two cases, they always return the same values for me:

```python
>>> f = lambda x: x**3 - 2.
>>> x1, x2 = 1.3, 1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
(1.3, 0.19700000000000006, 1.5, 1.375)

# -> abs(f1) < abs(f2)

>>> (f1*x2 - f2*x1) / (f1 - f2)
1.266553480475382

>>> (f2*x1 - f1*x2) / (f2-f1)
1.266553480475382

# both conditions return the same "x"

>>> x1, x2 = -1.3, 1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
>>> (-1.3, -4.197, 1.5, 1.375)

# -> abs(f1) > abs(f2)

>>> (f1*x2 - f2*x1) / (f1 - f2)
0.8090452261306534

>>> (f2*x1 - f1*x2) / (f2-f1)
0.8090452261306534

# both conditions return the same "x"

>>> x1, x2 = 1.3, -1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
(1.3, 0.19700000000000006, -1.5, -5.375)

# -> abs(f1) < abs(f2)

>>> (f1*x2 - f2*x1) / (f1 - f2)
1.2010050251256281

>>> (f2*x1 - f1*x2) / (f2-f1)
1.2010050251256281

# both conditions return the same "x"

>>> x1, x2 = -1.3, -1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
(-1.3, -4.197, -1.5, -5.375)

# -> abs(f2) > abs(f1)

>>> (f1*x2 - f2*x1) / (f1 - f2)
-0.5874363327674018

>>> (f2*x1 - f1*x2) / (f2-f1)
-0.5874363327674018

# both conditions return the same "x"
```

is this missing `sign()`? I remember you had a comment about the slope going the wrong way, but now I can't find it anywhere.

Reply:
Yes.  Both expressions attempt to compute the mathematical `(q1 p0 - q0 p1)/(q1 - q0)`. If `|q1/q0| > 1e16` then the contribution of `q0` can be lost.  So start from the endpoint with the smallest value of the function and move a slight distance away `p0 + epsilon`, rather than from the other end where one should really move `(p0-p1)+epsilon`, but `epsilon` is swamped by `p0-p1`. [Yes, some test cases really do have variation this large.]